### ISSUE TITLE
Managing combinatorial explosion: Limit number of returned results

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#2694]
We discussed that the ideal way to manage combinatorial explosion in the number of results is to implement "iterator interfaces". Doing this is a long way off. For igraph 1.0, I propose a simpler approach that is easy enough to implement, and can be practical for many applications: 

_Limit the number of results._

For example, a clique finder function currently lists _all_ cliques within the provided size range. If we only want one such clique, or a few, the only option is to use the callback version. Many of the callback interfaces are not exposed in high-level languages, as doing so is brittle and difficult. Even when they are, repeatedly calling a host-language function can cause slowdowns.

I propose that the following functions should receive an extra parameter which limits the number of results to return, making it easy to get one or a few results from high-level languages.

 - `igraph_cliques()`
 - `igraph_maximal_cliques()`
 - `igraph_largest_cliques()` *
 - `igraph_independent_vertex_sets()`
 - `igraph_weighted_cliques()`
 - `igraph_largest_weighted_cliques()` *
 - `igraph_maximal_independent_vertex_sets()`
 - `igraph_largest_independent_vertex_sets()` *
 - `igraph_get_all_simple_paths()`
 - `igraph_simple_cycles()`
 - `igraph_get_isomorphisms_vf2()`
 - `igraph_get_subisomorphisms_vf2()`
 - `igraph_subisomorphic_lad()` *

All of these list objects of which there can be exponentially many.

Perhaps the ones with a * don't need this, as there aren't usually that many results, or the implementation is not as straightforward as for the rest.

**Comment by vtraag** [comment]
Instead of adding an argument to these functions, could it perhaps be handled more gracefully by simply returning whatever results are already found when the function is interrupted? This also does not necessitate any API changes, so could be implemented after 1.0.

**Comment by szhorvat** [comment]
I don't agree that this approach will be useful, but I acknowledge the practical time constraints. Please leave this to me. What I can do before the release deadline will be in 1.0, the rest left for 2.0.

The real solution IMO is to (1) move to C++ first (2) implement iterators using C++20's coroutines.

**Comment by ntamas** [comment]
> Instead of adding an argument to these functions, could it perhaps be handled more gracefully by simply returning whatever results are already found when the function is interrupted? 

From the perspective of the C core, this would mean that functions supporting this kind of "graceful interruption" would need to guarantee that the output parameter in which the results are collected is in a valid state when the function returns `IGRAPH_INTERRUPTED`. I think this can be done. However, from the perspective of the maintainer of the Python interface, this would still not solve the problem. Let me elaborate why.

In Python, an interruption request (i.e. Ctrl-C from the user) is turned into a `KeyboardInterrupt` exception. We check for this  regularly in our interrupt handler and make the igraph function that was called return an `IGRAPH_INTERRUPTED` error code, but we do _not_ clear the exception at the moment - we let it propagate back to the caller. This means that after an interruption the user is not able to retrieve the result - Python functions either return successfully (with a value) or raise an exception, you can't have both. We would need to swallow `KeyboardInterrupt` proactively if we just wanted to let the user access the result after pressing Ctrl-C.

However, now, imagine a long-running server process that uses igraph to do some calculations in response to some requests. A request comes in, igraph starts calculating something that takes a long time, and then the operator comes and tries to shut the server down with a `SIGINT` (which is also turned into a `KeyboardInterrupt`). We will then happily capture the `KeyboardInterrupt`, pretend that it never happened, and the server process keeps on running, returning a partial result to the user. This is not what we want.

As @szhorvat mentioned, the "proper" solution would be C++20 coroutines in the C core, but I'm unsure how these could be converted into Python iterators. Maybe they could not be converted at all, in which case we could have fancy iterators in the C core that we are not able to exploit from Python. All in all, these possibilities require a deeper investigation and I don't think we have enough resources to do it properly before 1.0.

I'll just let @szhorvat do what he wants to be done for 1.0 but I won't consider this a blocker for the 1.0 release if it is not ready by then -- we need to draw a line somewhere. I would like to have _at least_ a release candidate out before 1 September.

**Comment by szhorvat** [comment]
> but I won't consider this a blocker for the 1.0 release if it is not ready by then -- we need to draw a line somewhere.

Yes, we are in agreement.

**Comment by ntamas** [comment]
@szhorvat Shall I merge #2824 before the release candidate (due the day after tomorrow) or shall we leave it and then do things "properly" after the release of 1.0?

**Comment by ntamas** [comment]
API changes done before the release of 1.0, implementation can remain for later, removing from 1.0 milestone for the time being.

**Comment by szhorvat** [comment]
OK, I would have also been happy to leave it as-is, and I completed all the high-priority items.

**Comment by szhorvat** [comment]
> API changes done before the release of 1.0, implementation can remain for later, removing from 1.0 milestone for the time being.

I thought you meant that you made the remaining API changes, but I do not see the commits for this. If you haven't made the API changes yet, I suggest NOT to do it. It is fine as it is. 

The implementations are actually completed, it's not just the API changes.
